### PR TITLE
🎨 Convert JDBC Importer to View Framework and Refactor importing

### PR DESCRIPTION
**Pull request by aldebaran30701**
### Description of the Change
This change is similar to the one done to the Delimited File Importer and has been done to increase view consistency.
This PR doesn't include a fix to the usability issues mentioned in another ticket. 
The PR includes extracting commonly used files into an inheritance hierarchy to cut down on code duplication. This should make it easier to extract the base steps of importing into it's own plugin in the future.

To note: This PR does not include a complete rework of the help documentation. I believe it will be best to do so when the workflow of the view changes as per @cygnus-x-1  review comments.

### Alternate Designs
N/A

### Why Should This Be In Core?
Consistency, and removes a heap of duplicated code through inheritance

### Benefits
Framework is consistent and user friendly.
Will be easier to create another importer which follows the same framework

### Possible Drawbacks
N/A

### Verification Process
Tested opening view via welcome page and also via the toolbar item

### Applicable Issues
#956

### CODE REVIEW

**Comment by aldebaran30701** [comment]
I've went through all of the Sonar warnings. 
The SQL vulnerability I couldn't manage to fix - switching variables and making the term final didn't seem to effect the net result.
Blockers like the "add a header to this file" seem to be possibly related to the 2020 copyright on my branch. 
Cleared all critical smells
Many high priority smells like the log or throw exception are false positives - We log them using LOGGER calls.
Other things like comment density | Split methods | lambdas too long were ignored.

**Comment by Auriga2** [comment]
A minor issue- When the `Manage  Connections` window is open, the user cannot open the help page. If they need to see the help page, it needs to be opened before the `Manage  Connections` window. Is there a way to improve this?
Also when the `Manage  Connections` window is open, if the user clicks on another View, the `Manage  Connections` window hides behind the screen, and locks some controls in the other views, making Constellation to appear stuck.

**Comment by aldebaran30701** [comment]
> A minor issue- When the `Manage Connections` window is open, the user cannot open the help page. If they need to see the help page, it needs to be opened before the `Manage Connections` window. Is there a way to improve this?
> Also when the `Manage Connections` window is open, if the user clicks on another View, the `Manage Connections` window hides behind the screen, and locks some controls in the other views, making Constellation to appear stuck.

Good spotting!
I've made it to now show as the front-most window while opened. This will allow it to never get stuck behind the main window which I feel is the root cause of the confusion when clicking other views. 
The only other alternate fix is to create a similar pane for managing the connections. ie. allow it to be done from the main view rather than the other dialog.

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/ConfigurationPane.java [hunk 1]
```diff
@@ -157,18 +156,15 @@ public void createNewRun(final Map<String, Attribute> vertexAttributes, final Ma
...
      *
      * @param columnLabels Column header labels.
      * @param currentData Rows of sample data.
      */
     public void setSampleData(final String[] columnLabels, final List<String[]> currentData) {
-        for (final Tab tab : tabPane.getTabs()) {
-            final RunPane runPane = (RunPane) tab.getContent();
+        tabPane.getTabs().stream().map(tab -> (RunPane) tab.getContent()).forEachOrdered(runPane -> {
             runPane.setSampleData(columnLabels, createTableRows(currentData));
-            runPane.refreshDataView();
```
Comment (Auriga2) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/ConfigurationPane.java:170]: This has reverted the change from https://github.com/constellation-app/constellation/pull/783/commits/e3c9eef8de696e268acbea846aab10a8dca37008 hence now I can see the mirror image on the Delimited File Importer- Preview table.

Comment (aldebaran30701) [reply CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/ConfigurationPane.java:170]: Good spotting!

**Comment by Auriga2** [comment]
It looks much better, but if the connection cannot be added, or test fails, the pop up error message appears behind these windows, which locks these windows until the user is responded on the prompt message. But that's not visible and only way to get away with this is by using the keyboard.

**Comment by aldebaran30701** [comment]
>It looks much better, but if the connection cannot be added, or test fails, the pop up error message appears behind these windows, which locks these windows until the user is responded on the prompt message. But that's not visible and only way to get away with this is by using the keyboard.

I couldn't get it to do as you mention - I couldn't get any popups to occur but that may be as a result of not knowing how to properly import a driver.

What I've done is remove the `dialog.setAlwaysOnTop(true);` call. I suspect that `dialog.toFront();` will be suffice to fix the issue of the manage connections window behind, and also fix what you describe.

**Comment by Auriga2** [comment]
Yes you need a driver to get anything to pop up. I used https://dev.mysql.com/downloads/connector/j/ (Download it and select the Jar file as the `Driver`. It populates the `Name` drop down.)

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/jdbc/JDBCSourcePane.java [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+                        connectionsTable.getItems().addAll(connectionManager.getConnections());
+                        connections.clear();
+                        connections.addAll(connectionManager.getConnections());
+                    }
+                });
+                connectionsPane.add(removeBtn, 0, 1, 1, 1);
+
+                final EventHandler<ActionEvent> addConnectionAction = new EventHandler<ActionEvent>() {
+                    @Override
+                    public void handle(final ActionEvent t1) {
```
Comment (antares1470) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/jdbc/JDBCSourcePane.java:162]: lambda if possible

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/jdbc/JDBCSourcePane.java [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+                                d.close();
+                            }
+                        });
+
+                        gp.add(add, 0, 5, 1, 1);
+
+                        final Button test = new Button("Test");
+                        test.setOnAction(new EventHandler<ActionEvent>() {
+                            @Override
+                            public void handle(final ActionEvent t2) {
```
Comment (antares1470) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/jdbc/JDBCSourcePane.java:218]: lambda if possible

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/jdbc/JDBCSourcePane.java [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+                driverTable.getColumns().addAll(name, jar);
+
+                driverTable.getItems().addAll(driverManager.getDrivers());
+                driverTable.getSortOrder().add(name);
+
+                dtRoot.add(driverTable, 0, 0, 2, 1);
+
+                final EventHandler<ActionEvent> removebtnAction = new EventHandler<ActionEvent>() {
+                    @Override
+                    public void handle(final ActionEvent t1) {
```
Comment (antares1470) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/jdbc/JDBCSourcePane.java:288]: lambda

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/jdbc/JDBCSourcePane.java [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+
+                    final Label jarLabel = new Label("Driver");
+                    gp.add(jarLabel, 0, 1, 1, 1);
+
+                    final TextField j = new TextField();
+                    gp.add(j, 1, 1, 1, 1);
+                    final Button chooser = new Button("..");
+                    chooser.setOnAction(new EventHandler<ActionEvent>() {
+                        @Override
+                        public void handle(final ActionEvent t2) {
```
Comment (antares1470) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/jdbc/JDBCSourcePane.java:346]: lambda

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/jdbc/JDBCSourcePane.java [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+                                }
+                            }
+                        }
+                    });
+                    gp.add(chooser, 2, 1, 1, 1);
+
+                    final Button add = new Button("Add");
+                    add.setOnAction(new EventHandler<ActionEvent>() {
+                        @Override
+                        public void handle(final ActionEvent t2) {
```
Comment (antares1470) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/jdbc/JDBCSourcePane.java:366]: lambda

**Review by antares1470 — CHANGES_REQUESTED** [review]

**Review by arcturus2 — CHANGES_REQUESTED** [review]
@aldebaran30701 can you please merge in the latest changes from `master` and advice it is ready for another review. Really keen to get this into the new release also.

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/ImportTableColumn.java [hunk 7]
```diff
@@ -72,35 +72,19 @@ public void validate(final List<TableRow> data) {
             final PluginParameters parserParameters = attributeNode.getTranslatorParameters();
             final String defaultValue = attributeNode.getDefaultValue();
 
-            Class<? extends AttributeDescription> attributeDescriptionClass = AttributeRegistry.getDefault().getAttributes().get(attributeNode.getAttribute().getAttributeType());
+            Class<? extends AttributeDescription> attributeDescriptionClass = AttributeRegistry.getDefault()
```
[the remark this answers is not in the document]
Comment (aldebaran30701) [reply CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/ImportTableColumn.java:75]: Line 79 sets the value if null so I can't make it final.

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/RunPane.java [hunk 8]
```diff
@@ -231,6 +234,8 @@ public RunPane(final ImportController importController) {
                 attributeFilter += c.getChar();
                 attributeFilterTextField.setText(attributeFilter);
                 attributeFilterPane.setVisible(true);
+            } else {
+                return; // Default case added per Sonar - java:S126
```
[the remark this answers is not in the document]
Comment (aldebaran30701) [reply CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/RunPane.java:238]: Not sure what went through my head when I was doing this 😆

**Comment by aldebaran30701** [comment]
@arcturus2  @antares1470  ready for review 👍

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/ImportTableColumn.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (antares1470) [reply CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/ImportTableColumn.java:75]: ah. never mind then

**Review by arcturus2 — APPROVED** [review]

**Review by antares1470 — CHANGES_REQUESTED** [review]
Gave a quick test and it's looking good. The only suggestion I have is aligning the css of the Options button to that of the Delimited File importer for consistency. There also appears to be a merge conflict.

**Review by antares1470 — APPROVED** [review]

**Pull request merged**